Issue 661 add geojson export type

ggalmazor opened the pull request:
Closes #661

Use this JAR if you want to try this one: 
[briefcase-669-geojson.jar.zip](https://github.com/opendatakit/briefcase/files/2565216/briefcase-669-geojson.jar.zip)

#### What has been done to verify that this works as intended?
Added several tests and run manually from the UI with forms that contained spatial fields and forms that didn't.

#### Why is this the best possible solution? Were any other approaches considered?
This PR uses the CSV export code as a scaffolding to build the GeoJSON export type. I aimed to do the least amount of changes, but I had to refactor the export conf class to make it immutable and make improvements around object validity which made following work easier to address and test.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Users enabling this new feature should expect an extra exported file with the same filename as the main CSV file, but with the `.geojson` extension. GeoJSON files can be opened in GIS software like QGIS.

To validate that the exported GeoJSON file is valid, they can be loaded into an online linter such as http://geojson.io

**UI - Including the GeoJSON companion export file** 

A new overridable boolean conf param has been added. This is how it works on the UI

![image](https://user-images.githubusercontent.com/205913/48252605-7e50e400-e405-11e8-911f-2fbdbfa85aa7.png)
(default conf panel)

![image](https://user-images.githubusercontent.com/205913/48252642-90cb1d80-e405-11e8-8084-82a44a76c50e.png)
(custom conf panel)

**CLI - Including the GeoJSON companion export file** 

There's a new CLI `-ig` `--include_geojson` flag users can use to enable including the GeoJSON export companion file.

```
Params for -e operation:
  -ed,--export_directory <arg>        Export directory
  -f,--export_filename <arg>          Filename for export operation
  -id,--form_id <arg>                 Form ID
  -sd,--storage_directory <arg>       Briefcase storage directory
Optional params for -e operation:
  -em,--exclude_media_export          Exclude media in export
  -end,--export_end_date <arg>        Export end date (inclusive) (yyyy-MM-dd or yyyy/MM/dd)
  -ig,--include_geojson               Include a GeoJSON companion file with spatial data
  -oc,--overwrite_csv_export          Overwrite files during export
  -pb,--pull_before                   Pull before export
  -pf,--pem_file <arg>                PEM file for form decryption
  -ssm,--split_select_multiples       Split select multiple fields
  -start,--export_start_date <arg>    Export start date (inclusive) (yyyy-MM-dd or yyyy/MM/dd)
```

**Changes to the export configuration**
This PR introduces a new invariant system that will check the validity of values introduced by the user to prevent launching exports with wrong value e.g. wrong pem file, wrong export directory, etc.

This system is used on the export dir and pem file conf params. To test this, try to set a wrong pem file. You will get an error dialog explaining what's the problem, and the field will be automatically cleared for you.

#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.
Yes, we should document the new CLI param. https://github.com/opendatakit/docs/issues/894

lognaturel commented:
Went through commit by commit and only had a couple of very surface suggestions.

There are a few things that I want to highlight:
- The spec and issue have the export type as a dropdown in the default configuration. Now, it's a radio button on the export tab:
<img width="498" alt="screen shot 2018-11-06 at 7 36 50 pm" src="https://user-images.githubusercontent.com/967540/48109214-5db53e00-e1fb-11e8-84de-a253c0bdf5f4.png">
I would prefer the dropdown because it's more consistent and leaves the door open for more export options. What's the reasoning for the radio buttons?

- The export type can't be set on a per-form basis. Was this a deliberate choice?

- The CLI feature has long name `--export_type` and short name `-et`. The description says `Export type: csv (default) or geojson` I would like to suggest a short name of `-t` but the rest looks good.

- The new invariants introduced will mean some users won't be able to launch Briefcase. For example, I can set a PEM file, delete it, try to launch Briefcase and then I get a stack trace with `Given PEM file doesn't exist`. It's possible to recover by clearing preferences but I am not clear on the advantage of checking these things before an export is attempted. If it's done at export time, the user can recover.

ggalmazor commented on src/org/opendatakit/briefcase/reused/Consumer3.java:
```diff
@@ -0,0 +1,22 @@
+package org.opendatakit.briefcase.reused;
+
+@FunctionalInterface
+public interface Consumer3<I1, I2, I3> {
```
The only problem is that it doesn't scale up very well (next would be Tetra?), but it doesn't seem likely we will end up with that, and I agree it's more consistent with Java's naming pattern

ggalmazor commented on src/org/opendatakit/briefcase/export/ExportType.java:
```diff
@@ -16,20 +16,27 @@
-  CSV(ExportToCsv::export),
-  GEOJSON(ExportToGeoJson::export);
+  CSV(ExportToCsv::export, ExportToCsv::export),
+  GEOJSON(ExportToGeoJson::export, ExportToGeoJson::export);
```
I think it's nicer to have methods that deal with `Optional` for you, but I agree the construction of these enums is awkward. I'll make the factory that accepts `Optional<Analytics>` package-private to inject only one factory.

ggalmazor commented:
@lognaturel 

You're right! I changed the dropdown without consulting it, sorry! Here's my rationale:

**Why radio buttons:**

We have a precedent similar change from a dropdown to radio buttons in the "overridable" booleans from the custom confs. 
  - They're easier to use. They show all the options in one glance
  - They're less bulky

I think it's OK to have them while everything fits nicely in one row (we could fit 4-ish options), and switching to a dropdown is not a big deal

I agree that the export type should be on a per-form basis, but here's **why it's currently outside the conf:**

- It felt like it was the smallest first deliverable increment to have people use this feature and get feedback before making it more complex.
  - This conf param seems special, more important than the rest
  - Having this option inside the default/custom confs should probably require more UI work i.e. making obvious directly in the forms table which format you've chosen for each selected form.

Regarding invariants, I think you're right and this could lead to some nasty problems for end users. I'll review it and try to maintain it (validity of value objects is an important constraint) and avoid errors while launching the app.

ggalmazor commented:
@lognaturel, I've changed the way we build ExportConfiguration objects to ensure it won't break when launching Briefcase. Let me know what you think :)

lognaturel commented on src/org/opendatakit/briefcase/operations/Export.java:
```diff
@@ -49,6 +49,7 @@
 public class Export {
   private static final Logger log = LoggerFactory.getLogger(Export.class);
   private static final Param<Void> EXPORT = Param.flag("e", "export", "Export a form");
+  private static final Param<ExportType> EXPORT_TYPE = Param.arg("et", "export_type", "Export type: csv (default) or geojson", ExportType::from);
```
Hehehe. Yes, I also thought that.

ggalmazor replied:
I've thrown in the commit https://github.com/opendatakit/briefcase/pull/669/commits/fbab22078b7ded062338adc4b58089dbabe9a280 that makes this change. We can skip it to maintain `-et` or fixup it to leave it as `-t`

lognaturel commented:
The invariant behavior on first launch looks good! I think it's totally fine for it to silently fail and log.

When the user does something that violates an invariant in normal usage, though, there should be UI feedback. This can be addressed in another issue if you prefer (like you could reopen #676 😬).

Thanks for the explanation on the UI changes. I forget that Briefcase users are fairly resilient to UI changes so doing a first pass like this and possibly changing it later does sound ok. Maybe @yanokwa wants to provide some feedback?

ggalmazor commented:
> When the user does something that violates an invariant in normal usage, though, there should be UI feedback. This can be addressed in another issue if you prefer (like you could reopen #676 grimacing).

The nice thing about this is that the UI won't let you press OK until the conf is valid, which prevents the issue in the first place. In any case, yes: we should reopen #676 and make a more general review of how feedback is given to users when something goes wrong.

ggalmazor commented:
> Thanks for the explanation on the UI changes. I forget that Briefcase users are fairly resilient to UI changes so doing a first pass like this and possibly changing it later does sound ok. Maybe @yanokwa wants to provide some feedback?

I'm happy that it sounds OK to you :) I wanted to avoid mixing two important conversations i.e. getting the geojson right is one thing and improving the UI is another thing. Let's hear what @yanokwa has to say :)

lognaturel commented:
> The nice thing about this is that the UI won't let you press OK until the conf is valid, which prevents the issue in the first place. 

That would be fine, too! For me it lets me click OK, it looks like the file is saved, but when I reopen the configuration there's nothing listed there. Then I get the stack trace for the issue every time I run an export.

ggalmazor commented:
> For me it lets me click OK, it looks like the file is saved, but when I reopen the configuration there's nothing listed there. Then I get the stack trace for the issue every time I run an export.

I will have to look into that.

ggalmazor commented:
@lognaturel, I've added a couple of commits covering user feedback to prevent the behavior you've detected. I've added a section to the PR description explaining this change on user-facing behavior ;)

ggalmazor commented:
Apologies! The good news is that it's simple to implement because all the necessary stuff is already there.

ggalmazor commented:
I've rebased master to solve conflicts and now this branch should be ready :)

dcbriccetti commented on src/org/opendatakit/briefcase/ui/export/components/ConfigurationPanelForm.java:
```diff
@@ -292,13 +293,20 @@ private void setPullBefore(OverridableBoolean value) {
-    overwriteFiles = overwriteFiles.overrideWith(overrideValue);
-    triggerOnChange();
+    TriStateBoolean previousValue = overwriteFiles.getOverride();
+    if (previousValue != overrideValue && (overrideValue != TRUE || confirmOverwriteFiles())) {
```
Not a big deal, but the all-caps TRUE threw me off until I found out what it is. No need to change it for me, but alternatives that occur to me:
overrideValue != TriStateBoolean.TRUE
! isTrue(overrideValue)

ggalmazor replied:
Thanks, Dave!

Maybe we could think of another enum name to use instead of `TRUE`?

kkrawczyk123 commented:
**Tested with success!**
Verified on Ubuntu, MacOS and Windows.

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"
